sw/host/opentitanlib/src/transport/verilator/subprocess.rs

```diff
@@ -0,0 +1,96 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::util::file;
+use anyhow::Result;
+use log::info;
+use regex::Regex;
+use std::io::Read;
+use std::process::{Child, ChildStdout, Command, Stdio};
+use std::time::{Duration, Instant};
+
+/// Verilator startup options.
+pub struct Options {
+    /// The verilator executable.
+    pub executable: String,
+    /// The ROM image used to boot the CPU.
+    pub rom_image: String,
+    /// The flash image stored in internal flash memory.
+    pub flash_image: String,
+    /// The OTP settings.
+    pub otp_image: String,
+    /// Any extra arguments to verilator.
+    pub extra_args: Vec<String>,
+}
+
+pub struct Subprocess {
+    child: Child,
+    stdout: ChildStdout,
+    accumulated_output: String,
+}
+
+impl Subprocess {
+    /// Starts a verilator [`Subprocess`] based on [`Options`].
+    pub fn from_options(options: Options) -> Result<Self> {
+        let mut command = Command::new(&options.executable);
+        let mut args = Vec::new();
+
+        if !options.rom_image.is_empty() {
+            args.push(format!("--meminit=rom,{}", options.rom_image));
+        }
+        if !options.flash_image.is_empty() {
+            args.push(format!("--meminit=flash,{}", options.flash_image));
+        }
+        if !options.otp_image.is_empty() {
+            args.push(format!("--meminit=otp,{}", options.otp_image));
+        }
+        args.extend_from_slice(&options.extra_args[..]);
+        command.args(&args[..]);
+
+        info!("Spawning verilator: {:?} {:?}", options.executable, args);
+
+        let mut child = command
+            .stdin(Stdio::null())
+            .stdout(Stdio::piped())
+            .stderr(Stdio::inherit())
+            .spawn()?;
+        let stdout = child.stdout.take().unwrap();
+        Ok(Subprocess {
+            child,
+            stdout,
+            accumulated_output: String::default(),
+        })
+    }
+
+    /// Accumulates output from verilator's `stdout`.
+    pub fn accumulate(&mut self, timeout: Duration) -> Result<()> {
+        let mut buf = [0u8; 256];
```

Comment:
I would say that since this is running on a machine with an allocator, allocating this much memory on the stack is slightly rude. I'd use a vector.

Reply:
I have the opposite reaction:  `buf` is tiny by comparison to a host machine's typical stack allocation (256 bytes vs. megabytes) and `buf` is a temporary which is used, converted, copied and then destroyed.  Asking the allocator to deal with it represents a source of memory fragmentation.

Considering that the contents of `buf` are converted to an utf8 string (ie: converted to a vec<u8>) and then appended onto another String, this discussion is somewhat academic: I'm already asking the allocator to allocate small chunks.

In general, for host-side code, stack allocations of up to a few kilobytes are not controversial:
https://github.com/google/boringssl/blob/master/crypto/pem/pem_info.c#L284  (1k buffer on stack)
https://github.com/google/boringssl/blob/master/crypto/err/err.c#L590 (1k + 120 bytes on stack)
(etc).

Reply:
Hehe, fair enough. =)